File: core/src/main/java/org/elkd/core/cluster/Node.java
package org.elkd.core.cluster;

import com.google.common.base.Preconditions;
import org.elkd.shared.schemes.URI;

import java.util.Objects;

public class Node {
  private URI mUri;

  public Node(final URI uri) {
    mUri = Preconditions.checkNotNull(uri, "uri");
  }

  public String getId() {
    return String.valueOf(mUri.hashCode());
  }

  public URI getURI() {
    return mUri;
  }

  public String getHost() {
    return mUri.getHost();
  }

  public int getPort() {
    return mUri.getPort();
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    final Node node = (Node) o;
    return Objects.equals(mUri, node.mUri);
  }

  @Override
  public int hashCode() {
    return Objects.hash(mUri);
  }

  @Override
  public String toString() {
    return mUri.toString();
  }
}

File: core/src/main/java/org/kerala/core/server/cluster/Node.kt
package org.kerala.core.server.cluster

import java.util.Objects
import org.kerala.shared.annotations.Mockable
import org.kerala.shared.schemes.URI

@Mockable
class Node(val uri: URI) {
  val id: String
    get() = uri.toString()

  val host: String
    get() = uri.host

  val port: Int
    get() = uri.port

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val node = other as Node?
    return uri == node!!.uri
  }

  override fun hashCode(): Int {
    return Objects.hash(uri)
  }

  override fun toString(): String {
    return id
  }
}
